### PR TITLE
speed up system flush logs

### PR DESCRIPTION
**Pull request by CheSema**
It was discovered here https://github.com/ClickHouse/ClickHouse/issues/67217#issuecomment-2256171604 that `SYSTEM FLUSH LOGS` takes almost a minute.

That operation flushes 20 system tables. It does it sequentially. Table by table. It leads to the situation when we waiting the data is flushed which is too fresh for us. And obviously it is one-thread process.

Im changing it into this:
- firstly the last log indexes are collected for all system tables, that what we need to flush, no more
- then for all system tables we trigger flush threads to start the work simultaneously in the background
- finally we wait until logs is flushed for all system tables.

There was no information how many logs are ignored when queue is full. Just a message that it had happened.
Was:
```
2024.07.31 19:12:40.127541 [ 759 ] {09f94353-43d1-47cc-a683-af2a17cdb9e0} <Error> SystemLogQueue (system.query_log): Queue is full for system log 'DB::SystemLogQueue<DB::QueryLogElement>' at 0. max_size_rows 10
```
Now:
```
2024.07.31 20:50:38.904990 [ 1419 ] {} <Error> SystemLog (system.query_log): Queue had been full at 0, accepted 10 logs, ignored 34 logs.
```

### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
speed up system flush logs, flush logs on shutdown.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [x]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

File: src/Common/SystemLogBase.cpp [hunk 1]
```diff
@@ -27,12 +27,15 @@
 #include <Common/logger_useful.h>
 #include <base/scope_guard.h>
 
+#define LOGICAL_IF_THEN(A, B) (!(A) || !!(B))
```
Comment (nikitamikhaylov) [inline src/Common/SystemLogBase.cpp:30]: Move to `base/defines.h` ?

Comment (CheSema) [reply src/Common/SystemLogBase.cpp:30]: Ok, I don't mind.

Comment (Algunenano) [reply src/Common/SystemLogBase.cpp:30]: I'm not sure adding this to defines.h, which is included everywhere and will force recompiling everything, is a good idea. It's a macro only used once, so maybe we could even not use a macro.

Comment (CheSema) [reply src/Common/SystemLogBase.cpp:30]: It is better to be a function in general.
For example if we do 
```
#define LOGICAL_IFF(A, B) (LOGICAL_IF_THEN(A, B) && LOGICAL_IF_THEN(B, A))
```
then `A` and `B` could be calculated twice.

But I wanted to leave it as simple as possible. Beside `LOGICAL_IFF` does not have such issue.

File: src/Backups/BackupsWorker.cpp [hunk 2]
```diff
@@ -1157,6 +1157,8 @@ void BackupsWorker::waitAll()
     for (const auto & id : current_operations)
         wait(id, /* rethrow_exception= */ false);
 
+    backup_log->flush(backup_log->getLastLogIndex());
```
Comment (CheSema) [inline src/Backups/BackupsWorker.cpp:1160]: There is some tests
`test_shutdown_cancel_backup` and `test_shutdown_wait_backup` they expects that before shutdown all the logs is flushed and after restart the tests see the result. 
I do not see where we do that in the code. Do we have such gurantee?

Comment (CheSema) [reply src/Backups/BackupsWorker.cpp:1160]: We could do the flush only for backups logs or make it for all table at `SystemLogs::shutdown`.

**Comment by CheSema** [comment]
Offtop: 
```
std::vector<ISystemLog *> Context::getSystemLogs() const
{
    SharedLockGuard lock(shared->mutex);

    if (!shared->system_logs)
        return {};
    return shared->system_logs->logs;
}
```
What the point of that method? It looks like we protected by mutex but in reality we could have data race and all that pointers are freed.

File: src/Backups/BackupsWorker.cpp (hunk 2 above)
Comment (CheSema) [reply src/Backups/BackupsWorker.cpp]: Will add flush for all logs at shutdown.

**Comment by CheSema** [comment]
Ready for review. Tests adjustments are left.

File: src/Interpreters/SystemLog.cpp [hunk 3]
```diff
@@ -462,41 +457,33 @@ void SystemLog<LogElement>::savingThreadFunction()
...
+            auto result = queue->pop();
 
-            if (to_flush.empty())
+            if (result.is_shutdown)
             {
-                if (should_prepare_tables_anyway)
-                {
-                    prepareTable();
-                    LOG_TRACE(log, "Table created (force)");
+                LOG_TRACE(log, "Terminating");
```
Comment (nikitamikhaylov) [inline src/Interpreters/SystemLog.cpp:468]: Should we include the table name here? Or does the logger name contain it already?

File: src/Interpreters/SystemLog.h [hunk 4]
```diff
@@ -5,6 +5,32 @@
 #include <Parsers/IAST.h>
 
 #include <boost/noncopyable.hpp>
+#include <vector>
+
+#define LIST_OF_ALL_SYSTEM_LOGS(M) \
+    M(QueryLog, query_log, "Used to log queries.") \
```
Comment (nikitamikhaylov) [inline src/Interpreters/SystemLog.h:11]: FYI We have comments here: https://github.com/ClickHouse/ClickHouse/blob/4cdb6db1b7bb212d655cde15c0eb8534687d6c9d/src/Interpreters/SystemLog.cpp#L285-L312

Maybe we can just reuse them?

**Review by nikitamikhaylov — APPROVED** [review]

File: src/Interpreters/SystemLog.h (hunk 4 above)
Comment (CheSema) [reply src/Interpreters/SystemLog.h:11]: Actually I do not need the description in LIST_OF_ALL_SYSTEM_LOGS. I just wanted to save comments from old code. 
But you are right, no need to do it twice, I will change it.

File: src/Interpreters/SystemLog.cpp (hunk 3 above)
Comment (CheSema) [reply src/Interpreters/SystemLog.cpp:468]: Yes, `log` are pretty good there, like `SystemLog (system.query_log)`.

**Comment by CheSema** [comment]
This is interesting:
4 tests have failed with `DB::Exception: Timeout exceeded (180 s) while flushing system log 'DB::SystemLogQueue<DB::TraceLogElement>'. (TIMEOUT_EXCEEDED)`.
There is definitely was a stopper.

CH triggered flush by size:
```
2024.08.08 09:52:31.862100 [ 897 ] {} <Information> SystemLogQueue (system.trace_log): Queue is half full for system log 'DB::SystemLogQueue<DB::TraceLogElement>'. buffer_size_rows_flush_threshold 524288
```

after i see a series of 
```
2024.08.08 09:52:42.683940 [ 1532 ] {} <Information> system.trace_log (08349c8e-1ac1-483c-9e61-e984973a099f): /var/lib/clickhouse/store/083/08349c8e-1ac1-483c-9e61-e984973a099f/tmp_insert_202408_250_250_0/ is in use (by merge/mutation/INSERT) (consider increasing temporary_directories_lifetime setting) (skipped 1 similar messages)
```
until
```
2024.08.08 09:53:16.830544 [ 1571 ] {} <Trace> system.trace_log (08349c8e-1ac1-483c-9e61-e984973a099f): Renaming temporary part tmp_insert_202408_250_250_0 to 202408_250_250_0 with tid (1, 1, 00000000-0000-0000-0000-000000000000).
```
A minute?

```
2024.08.08 09:53:16.831348 [ 1571 ] {} <Trace> SystemLog (system.trace_log): Flushed system log up to offset 23688188
```

```
2024.08.08 09:53:17.489717 [ 1571 ] {} <Error> SystemLogQueue (system.trace_log): Queue had been full at 23688188, accepted 1048576 logs, ignored 1765221 logs.
2024.08.08 09:53:17.489770 [ 1571 ] {} <Trace> SystemLog (system.trace_log): Flushing system log, 1048576 entries to flush up to offset 24736764.
```

Wow, we skip more that save.

```
2024.08.08 09:54:25.932434 [ 1456 ] {} <Error> system.trace_log_sender.DistributedInsertQueue.default: Code: 241. DB::Exception: Received from kng4alm55c.us-east-2.aws.clickhouse-staging.com:9440. DB::Exception: Memory limit (for user) exceeded: would use 354.13 GiB (attempt to allocate chunk of 16772415 bytes), maximum: 18.63 GiB. OvercommitTracker decision: Query was selected to stop by OvercommitTracker.. Stack trace:
```

And we have a troubles to send them.

**Pull request merged**